pkg/providers/tinkerbell/upgrade.go

```diff
@@ -506,3 +506,20 @@ func toRufioMachines(items []rufiounreleased.BaseboardManagement) []rufiov1.Mach
 	}
 	return machines
 }
+
+// IsRollingNotScale compares current CAPI objects and desired CAPI objects
+// returns true on rolling upgrade, false on scale upgrade or create
+// should error if rolling and scale at the same time
+// todo: IsRollingNotScale
+func IsRollingNotScale(currentState *ClusterApiState, clusterSpec *cluster.Spec) (bool, error) {
```

Comment:
I would suggest using an interface and comparing two of the same interface
That allows you to change the types you use to represent current and desired state. You are already using method to access the data, so that actually comes for free.